File: ruby/word-count/word_count.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/word-count
# Word Count exercise
class Phrase
  def initialize(text)
    @text = text.chomp.strip
  end

  def word_count
    @words = @text.scan(/\b[[:alpha:][:digit:]']+\b/).collect

    @counts = Hash.new(0)

    @words.each { |word| @counts[word.downcase] += 1 }

    @counts
  end
end

File: python/word-count/word_count.py
"""Word Count Python Exercism"""


import re
from collections import defaultdict
from typing import Dict


def count_words(sentence: str) -> Dict[str, int]:
    """Counts the number of times each word in a string is used.

    :param sentence: str - string we're going to count
    :return: Dict[str, int] - current round and the two that follow.
    """
    counts: Dict[str, int] = defaultdict(int)

    # remove surrounding quotes
    sentence = sentence.strip("'")

    # replace tabs with spaces
    sentence = sentence.expandtabs(tabsize=1)

    # replace punctuation and funny characters with spaces
    sentence = re.sub(r"[,.;:_!&@$%^&]", " ", sentence)

    for line in sentence.splitlines():
        for word in line.split(" "):
            # skip empty words
            if not word:
                continue

            # lower case the dict key
            word = word.lower()

            # remove surrounding quotes
            word = word.strip("'")

            # increment the word count in the dictionary
            counts[word] += 1

    return counts
